File: Engine/src/tungsten/docs_index.py
from __future__ import annotations

import json
import re
import sqlite3
import subprocess
from shutil import which
from dataclasses import dataclass
from pathlib import Path

from .discovery import (
    WolframInstallation,
    discover_installation,
    ensure_parent_directory,
    iter_notebook_files,
)
from .notebook import collapse_text, extract_string_literals


REFERENCE_CATEGORY_MAP = {
    "Symbols": "ref",
    "Programs": "ref/program",
    "MenuItems": "ref/menuitem",
    "Characters": "ref/character",
    "Entities": "ref/entity",
    "Interpreters": "ref/interpreter",
    "FrontEndObjects": "ref/frontendobject",
}

SECTION_CATEGORY_MAP = {
    "Guides": "guide",
    "Tutorials": "tutorial",
    "HowTos": "howto",
    "Workflows": "workflow",
    "WorkflowGuides": "workflowguide",
    "ExamplePages": "example",
}

NOISE_LITERALS = {
    "AnchorBar",
    "AnchorBarGrid",
    "Columns",
    "ExampleCount",
    "ExampleSection",
    "LinkHand",
    "ObjectNameTranslation",
    "PacletNameCell",
    "PrimaryExamplesSection",
    "Rows",
    "SeeAlsoRelated",
    "Spacer1",
}


@dataclass(frozen=True)
class DocumentationRecord:
    title: str
    paclet: str
    kind: str
    category: str
    path: str
    preview: str
    text: str

    def to_dict(self) -> dict[str, object]:
        return {
            "title": self.title,
            "paclet": self.paclet,
            "kind": self.kind,
            "category": self.category,
            "path": self.path,
            "preview": self.preview,
            "text": self.text,
        }


class DocumentationIndex:
    def __init__(self, installation: WolframInstallation | None = None) -> None:
        self.installation = installation or discover_installation()

    def ensure_index(self, index_path: Path | None = None, rebuild: bool = False) -> Path:
        target = index_path or self.installation.default_index_path
        if rebuild or not target.exists():
            self.build_index(target)
        return target

    def build_index(self, index_path: Path | None = None) -> Path:
        target = index_path or self.installation.default_index_path
        ensure_parent_directory(target)
        if target.exists():
            target.unlink()

        connection = sqlite3.connect(target)
        try:
            connection.execute(
                """
                CREATE TABLE documents (
                    id INTEGER PRIMARY KEY,
                    title TEXT NOT NULL,
                    paclet TEXT NOT NULL,
                    kind TEXT NOT NULL,
                    category TEXT NOT NULL,
                    path TEXT NOT NULL,
                    preview TEXT NOT NULL,
                    text TEXT NOT NULL
                )
                """
            )
            connection.execute(
                """
                CREATE VIRTUAL TABLE documents_fts USING fts5(
                    title,
                    paclet,
                    kind,
                    category,
                    preview,
                    text,
                    content='documents',
                    content_rowid='id'
                )
                """
            )
            connection.execute(
                """
                CREATE TABLE metadata (
                    key TEXT PRIMARY KEY,
                    value TEXT NOT NULL
                )
                """
            )

            roots = [str(root) for root in self.installation.docs_roots]
            connection.execute(
                "INSERT INTO metadata(key, value) VALUES(?, ?)",
                ("docs_roots", json.dumps(roots)),
            )

            for notebook_path in iter_notebook_files(self.installation.docs_roots):
                record = self._record_from_path(notebook_path)
                cursor = connection.execute(
                    """
                    INSERT INTO documents(title, paclet, kind, category, path, preview, text)
                    VALUES(?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        record.title,
                        record.paclet,
                        record.kind,
                        record.category,
                        record.path,
                        record.preview,
                        record.text,
                    ),
                )
                connection.execute(
                    """
                    INSERT INTO documents_fts(rowid, title, paclet, kind, category, preview, text)
                    VALUES(?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        cursor.lastrowid,
                        record.title,
                        record.paclet,
                        record.kind,
                        record.category,
                        record.preview,
                        record.text,
                    ),
                )

            connection.commit()
        finally:
            connection.close()

        return target

    def search(
        self,
        query: str,
        *,
        index_path: Path | None = None,
        limit: int = 10,
        rebuild: bool = False,
    ) -> list[dict[str, object]]:
        fast_hits = self._search_by_filename(query, limit=limit)
        if fast_hits:
            return fast_hits

        target = self.ensure_index(index_path=index_path, rebuild=rebuild)
        connection = sqlite3.connect(target)
        connection.row_factory = sqlite3.Row
        try:
            rows = connection.execute(
                """
                SELECT
                    documents.title,
                    documents.paclet,
                    documents.kind,
                    documents.category,
                    documents.path,
                    documents.preview,
                    snippet(documents_fts, 5, '[', ']', ' … ', 18) AS snippet,
                    bm25(documents_fts) AS score
                FROM documents_fts
                JOIN documents ON documents.id = documents_fts.rowid
                WHERE documents_fts MATCH ?
                ORDER BY score
                LIMIT ?
                """,
                (self._build_match_query(query), limit),
            ).fetchall()
        finally:
            connection.close()

        return [dict(row) for row in rows]

    def read(
        self,
        identifier: str,
        *,
        index_path: Path | None = None,
        rebuild: bool = False,
    ) -> dict[str, object]:
        fast_hits = self._search_by_filename(identifier, limit=1)
        if fast_hits:
            return fast_hits[0]

        target = self.ensure_index(index_path=index_path, rebuild=rebuild)
        connection = sqlite3.connect(target)
        connection.row_factory = sqlite3.Row
        try:
            if Path(identifier).exists():
                row = connection.execute(
                    "SELECT * FROM documents WHERE path = ?",
                    (str(Path(identifier).resolve()),),
                ).fetchone()
            elif identifier.startswith("paclet:"):
                row = connection.execute(
                    "SELECT * FROM documents WHERE paclet = ? COLLATE NOCASE",
                    (identifier,),
                ).fetchone()
            else:
                row = connection.execute(
                    """
                    SELECT * FROM documents
                    WHERE title = ? COLLATE NOCASE OR paclet = ? COLLATE NOCASE
                    LIMIT 1
                    """,
                    (identifier, identifier),
                ).fetchone()

                if row is None:
                    hits = self.search(identifier, index_path=target, limit=1)
                    if not hits:
                        raise ValueError(f"No documentation page found for {identifier!r}.")
                    row = connection.execute(
                        "SELECT * FROM documents WHERE paclet = ?",
                        (hits[0]["paclet"],),
                    ).fetchone()
        finally:
            connection.close()

        if row is None:
            raise ValueError(f"No documentation page found for {identifier!r}.")

        return dict(row)

    def resolve_identifier(
        self,
        identifier: str,
        *,
        index_path: Path | None = None,
    ) -> str:
        if identifier.startswith("paclet:"):
            return identifier
        return str(self.read(identifier, index_path=index_path)["paclet"])

    def _search_by_filename(self, query: str, *, limit: int) -> list[dict[str, object]]:
        stem = self._stem_from_identifier(query)
        if not stem:
            return []

        paths = self._find_notebook_paths(stem, limit=limit * 4)
        if not paths:
            return []

        records: list[dict[str, object]] = []
        for path in paths[:limit]:
            record = self._record_from_path(path)
            records.append(
                {
                    "title": record.title,
                    "paclet": record.paclet,
                    "kind": record.kind,
                    "category": record.category,
                    "path": record.path,
                    "preview": record.preview,
                    "snippet": record.preview,
                    "score": 0.0,
                }
            )
        return records

    def _record_from_path(self, notebook_path: Path) -> DocumentationRecord:
        raw = notebook_path.read_text(encoding="utf-8", errors="replace")
        title = self._extract_title(raw, notebook_path)
        kind, category, paclet = self._infer_kind_and_paclet(notebook_path)
        strings = self._filter_useful_strings(extract_string_literals(raw))
        text = collapse_text(" ".join(strings), limit=20_000)
        preview_source = " ".join(
            fragment for fragment in strings if fragment and fragment != title
        )
        preview = collapse_text(preview_source, limit=300)

        return DocumentationRecord(
            title=title,
            paclet=paclet,
            kind=kind,
            category=category,
            path=str(notebook_path.resolve()),
            preview=preview,
            text=text,
        )

    @staticmethod
    def _extract_title(raw: str, notebook_path: Path) -> str:
        match = re.search(r"WindowTitle->(?P<title>\"(?:\\\\.|[^\"])*\"|[A-Za-z0-9`.$_-]+)", raw)
        if match:
            title = match.group("title").strip()
            if title.startswith("\"") and title.endswith("\""):
                return title[1:-1].replace("\\\"", "\"")
            return title
        return notebook_path.stem

    @staticmethod
    def _infer_kind_and_paclet(notebook_path: Path) -> tuple[str, str, str]:
        parts = notebook_path.parts
        if "ReferencePages" in parts:
            category = parts[parts.index("ReferencePages") + 1]
            paclet_category = REFERENCE_CATEGORY_MAP.get(category, f"ref/{category.lower()}")
            return "reference", category, f"paclet:{paclet_category}/{notebook_path.stem}"

        for section, paclet_category in SECTION_CATEGORY_MAP.items():
            if section in parts:
                return paclet_category, section, f"paclet:{paclet_category}/{notebook_path.stem}"

        return "document", "Other", f"paclet:document/{notebook_path.stem}"

    @staticmethod
    def _build_match_query(query: str) -> str:
        terms = re.findall(r"[A-Za-z0-9_.:/-]+", query)
        if not terms:
            return f"\"{query}\""
        return " AND ".join(f"\"{term}\"*" for term in terms)

    @staticmethod
    def _filter_useful_strings(strings: list[str]) -> list[str]:
        uuid_pattern = re.compile(
            r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
        )
        compressed_blob = re.compile(r"^[A-Za-z0-9+/=:._-]{200,}$")

        filtered: list[str] = []
        for value in strings:
            trimmed = value.strip()
            if not trimmed:
                continue
            if trimmed in NOISE_LITERALS:
                continue
            if uuid_pattern.fullmatch(trimmed):
                continue
            if compressed_blob.fullmatch(trimmed):
                continue
            filtered.append(trimmed)
        return filtered

    @staticmethod
    def _stem_from_identifier(identifier: str) -> str:
        candidate = identifier
        if identifier.startswith("paclet:"):
            candidate = identifier.rsplit("/", 1)[-1]
        candidate = Path(candidate).stem
        if not candidate:
            return ""
        if not re.fullmatch(r"[A-Za-z0-9_.-]+", candidate):
            return ""
        return candidate

    def _find_notebook_paths(self, stem: str, *, limit: int) -> list[Path]:
        es_exe = which("es.exe")
        candidates: list[Path] = []
        roots = [root.resolve() for root in self.installation.docs_roots]

        if es_exe:
            completed = subprocess.run(
                [es_exe, f"{stem}.nb"],
                capture_output=True,
                text=True,
                encoding="utf-8",
                errors="replace",
                check=False,
            )
            for line in completed.stdout.splitlines():
                line = line.strip()
                if not line:
                    continue
                candidates.append(Path(line))
        else:
            for root in roots:
                candidates.extend(root.rglob(f"{stem}.nb"))

        filtered: list[Path] = []
        seen: set[Path] = set()
        for candidate in candidates:
            try:
                resolved = candidate.resolve()
            except OSError:
                continue

            if resolved in seen:
                continue
            if resolved.suffix.lower() != ".nb":
                continue
            if resolved.stem.lower() != stem.lower():
                continue
            if not any(str(resolved).lower().startswith(str(root).lower()) for root in roots):
                continue

            seen.add(resolved)
            filtered.append(resolved)

        filtered.sort(key=self._root_priority)
        return filtered[:limit]

    def _root_priority(self, path: Path) -> tuple[int, str]:
        normalized = str(path.resolve()).lower()
        for index, root in enumerate(self.installation.docs_roots):
            prefix = str(root.resolve()).lower()
            if normalized.startswith(prefix):
                return index, normalized
        return len(self.installation.docs_roots), normalized

File: Engine/haskell/src/Tungsten/DocsIndex.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Kernel-free extraction and SQLite FTS indexing of Wolfram documentation.
module Tungsten.DocsIndex
  ( DocumentationError (..)
  , DocumentationRecord (..)
  , DocumentationHit (..)
  , buildDocumentationIndex
  , ensureDocumentationIndex
  , searchDocumentation
  , readDocumentation
  , resolveDocumentationIdentifier
  , documentationRecordFromPath
  ) where

import Control.Exception (IOException, try)
import Control.Monad (forM, forM_)
import qualified Data.ByteString as BS
import Data.Char (isAlphaNum, isHexDigit, isSpace, toLower)
import Data.List (findIndex, sort)
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import qualified Data.Text.Encoding.Error as TEE
import qualified Data.Text.IO as TextIO
import Numeric (showHex)
import System.Directory
  ( createDirectoryIfMissing
  , doesDirectoryExist
  , doesFileExist
  , findExecutable
  , listDirectory
  , makeAbsolute
  , removeFile
  )
import System.Exit (ExitCode (..))
import System.FilePath
  ( (</>)
  , splitDirectories
  , takeBaseName
  , takeDirectory
  , takeExtension
  )
import System.IO (Handle, hClose, hSetEncoding, utf8)
import System.Process
  ( CreateProcess (..)
  , StdStream (CreatePipe, NoStream)
  , createProcess
  , proc
  , readProcessWithExitCode
  , waitForProcess
  )
import Text.Read (readMaybe)
import Tungsten.Discovery
import Tungsten.Json
import Tungsten.WolframString

newtype DocumentationError = DocumentationError {documentationErrorMessage :: Text}
  deriving (Eq, Show)

data DocumentationRecord = DocumentationRecord
  { documentationId :: !(Maybe Integer)
  , documentationTitle :: !Text
  , documentationPaclet :: !Text
  , documentationKind :: !Text
  , documentationCategory :: !Text
  , documentationPath :: !FilePath
  , documentationPreview :: !Text
  , documentationText :: !Text
  }
  deriving (Eq, Show)

data DocumentationHit = DocumentationHit
  { documentationHitTitle :: !Text
  , documentationHitPaclet :: !Text
  , documentationHitKind :: !Text
  , documentationHitCategory :: !Text
  , documentationHitPath :: !FilePath
  , documentationHitPreview :: !Text
  , documentationHitSnippet :: !Text
  , documentationHitScore :: !Text
  }
  deriving (Eq, Show)

buildDocumentationIndex
  :: WolframInstallation
  -> Maybe FilePath
  -> IO (Either DocumentationError FilePath)
buildDocumentationIndex installation requestedPath = do
  let target = maybe (installationDefaultIndexPath installation) id requestedPath
  absoluteTarget <- makeAbsolute target
  sqlite <- findExecutable "sqlite3"
  case sqlite of
    Nothing -> pure (Left (DocumentationError "sqlite3 is required to build the documentation index"))
    Just sqliteExecutable -> do
      preparation <- try $ do
        createDirectoryIfMissing True (takeDirectory absoluteTarget)
        exists <- doesFileExist absoluteTarget
        if exists then removeFile absoluteTarget else pure ()
      case (preparation :: Either IOException ()) of
        Left exception -> pure (Left (DocumentationError (T.pack (show exception))))
        Right () -> do
          files <- documentationNotebookFiles (installationDocsRoots installation)
          recordsResult <- traverse documentationRecordFromPath files
          case sequence recordsResult of
            Left documentationError -> pure (Left documentationError)
            Right records -> do
              writeResult <- writeSqliteIndex sqliteExecutable absoluteTarget installation records
              pure (absoluteTarget <$ writeResult)

ensureDocumentationIndex
  :: WolframInstallation
  -> Maybe FilePath
  -> Bool
  -> IO (Either DocumentationError FilePath)
ensureDocumentationIndex installation requestedPath rebuild = do
  let target = maybe (installationDefaultIndexPath installation) id requestedPath
  exists <- doesFileExist target
  if rebuild || not exists
    then buildDocumentationIndex installation requestedPath
    else Right <$> makeAbsolute target

searchDocumentation
  :: WolframInstallation
  -> Text
  -> Maybe FilePath
  -> Int
  -> Bool
  -> IO (Either DocumentationError [DocumentationHit])
searchDocumentation installation query indexPath limit rebuild = do
  fastPaths <- findDocumentationPaths installation query (max 0 limit)
  if null fastPaths
    then do
      targetResult <- ensureDocumentationIndex installation indexPath rebuild
      case targetResult of
        Left documentationError -> pure (Left documentationError)
        Right target -> searchSqlite target query (max 0 limit)
    else do
      records <- traverse documentationRecordFromPath fastPaths
      pure (map recordHit <$> sequence records)
 where
  recordHit record =
    DocumentationHit
      { documentationHitTitle = documentationTitle record
      , documentationHitPaclet = documentationPaclet record
      , documentationHitKind = documentationKind record
      , documentationHitCategory = documentationCategory record
      , documentationHitPath = documentationPath record
      , documentationHitPreview = documentationPreview record
      , documentationHitSnippet = documentationPreview record
      , documentationHitScore = "0.0"
      }

readDocumentation
  :: WolframInstallation
  -> Text
  -> Maybe FilePath
  -> Bool
  -> IO (Either DocumentationError DocumentationRecord)
readDocumentation installation identifier indexPath rebuild = do
  fastPaths <- findDocumentationPaths installation identifier 1
  case fastPaths of
    path : _ -> documentationRecordFromPath path
    [] -> do
      targetResult <- ensureDocumentationIndex installation indexPath rebuild
      case targetResult of
        Left documentationError -> pure (Left documentationError)
        Right target -> do
          primary <- readSqlite target identifier
          case primary of
            Right (Just record) -> pure (Right record)
            Left documentationError -> pure (Left documentationError)
            Right Nothing -> do
              hits <- searchDocumentation installation identifier (Just target) 1 False
              case hits of
                Right (hit : _) -> do
                  fallback <- readSqlite target (documentationHitPaclet hit)
                  pure $ case fallback of
                    Right (Just record) -> Right record
                    Right Nothing -> notFound identifier
                    Left documentationError -> Left documentationError
                Right [] -> pure (notFound identifier)
                Left documentationError -> pure (Left documentationError)

resolveDocumentationIdentifier
  :: WolframInstallation
  -> Text
  -> Maybe FilePath
  -> IO (Either DocumentationError Text)
resolveDocumentationIdentifier installation identifier indexPath
  | "paclet:" `T.isPrefixOf` identifier = pure (Right identifier)
  | otherwise =
      fmap documentationPaclet <$> readDocumentation installation identifier indexPath False

documentationRecordFromPath :: FilePath -> IO (Either DocumentationError DocumentationRecord)
documentationRecordFromPath notebookPath = do
  sourceResult <- try (BS.readFile notebookPath)
  case (sourceResult :: Either IOException BS.ByteString) of
    Left exception -> pure (Left (DocumentationError (T.pack (show exception))))
    Right bytes -> do
      absolutePath <- makeAbsolute notebookPath
      let source = TE.decodeUtf8With TEE.lenientDecode bytes
          title = extractTitle source notebookPath
          (kind, category, paclet) = inferKindAndPaclet notebookPath
          strings = filterUsefulStrings (extractStringLiterals source)
          textValue = collapseText 20000 (T.unwords strings)
          previewSource = T.unwords [fragment | fragment <- strings, fragment /= title]
      pure
        ( Right
            DocumentationRecord
              { documentationId = Nothing
              , documentationTitle = title
              , documentationPaclet = paclet
              , documentationKind = kind
              , documentationCategory = category
              , documentationPath = absolutePath
              , documentationPreview = collapseText 300 previewSource
              , documentationText = textValue
              }
        )

writeSqliteIndex
  :: FilePath
  -> FilePath
  -> WolframInstallation
  -> [DocumentationRecord]
  -> IO (Either DocumentationError ())
writeSqliteIndex sqlite target installation records = do
  processResult <- try $ do
    (Just input, _, Just errors, processHandle) <-
      createProcess
        (proc sqlite [target])
          { std_in = CreatePipe
          , std_out = NoStream
          , std_err = CreatePipe
          }
    hSetEncoding input utf8
    hSetEncoding errors utf8
    writeIndexScript input installation records
    hClose input
    errorText <- TextIO.hGetContents errors
    exitCode <- waitForProcess processHandle
    pure (exitCode, errorText)
  case (processResult :: Either IOException (ExitCode, Text)) of
    Left exception -> pure (Left (DocumentationError (T.pack (show exception))))
    Right (ExitSuccess, _) -> pure (Right ())
    Right (ExitFailure code, errorText) ->
      pure
        ( Left
            ( DocumentationError
                ("sqlite3 exited with code " <> T.pack (show code) <> ": " <> T.strip errorText)
            )
        )

writeIndexScript :: Handle -> WolframInstallation -> [DocumentationRecord] -> IO ()
writeIndexScript input installation records = do
  TextIO.hPutStrLn input ".bail on"
  TextIO.hPutStrLn input "BEGIN IMMEDIATE;"
  TextIO.hPutStrLn input
    "CREATE TABLE documents (id INTEGER PRIMARY KEY, title TEXT NOT NULL, paclet TEXT NOT NULL, kind TEXT NOT NULL, category TEXT NOT NULL, path TEXT NOT NULL, preview TEXT NOT NULL, text TEXT NOT NULL);"
  TextIO.hPutStrLn input
    "CREATE VIRTUAL TABLE documents_fts USING fts5(title, paclet, kind, category, preview, text, content='documents', content_rowid='id');"
  TextIO.hPutStrLn input "CREATE TABLE metadata (key TEXT PRIMARY KEY, value TEXT NOT NULL);"
  let rootsJson = encodeJson (JsonArray (map (JsonString . T.pack) (installationDocsRoots installation)))
  TextIO.hPutStrLn input
    ("INSERT INTO metadata(key, value) VALUES(" <> sqlText "docs_roots" <> ", " <> sqlText rootsJson <> ");")
  forM_ records $ \record ->
    TextIO.hPutStrLn input
      ( "INSERT INTO documents(title, paclet, kind, category, path, preview, text) VALUES("
          <> T.intercalate
            ", "
            ( map
                sqlText
                [ documentationTitle record
                , documentationPaclet record
                , documentationKind record
                , documentationCategory record
                , T.pack (documentationPath record)
                , documentationPreview record
                , documentationText record
                ]
            )
          <> ");"
      )
  TextIO.hPutStrLn input "INSERT INTO documents_fts(documents_fts) VALUES('rebuild');"
  TextIO.hPutStrLn input "COMMIT;"

searchSqlite :: FilePath -> Text -> Int -> IO (Either DocumentationError [DocumentationHit])
searchSqlite target query limit = do
  let statement =
        "SELECT documents.title, documents.paclet, documents.kind, documents.category, documents.path, documents.preview, "
          <> "snippet(documents_fts, 5, '[', ']', ' … ', 18) AS snippet, bm25(documents_fts) AS score "
          <> "FROM documents_fts JOIN documents ON documents.id = documents_fts.rowid "
          <> "WHERE documents_fts MATCH "
          <> sqlText (buildMatchQuery query)
          <> " ORDER BY score LIMIT "
          <> T.pack (show limit)
          <> ";"
  rows <- sqliteJsonQuery target statement
  pure (rows >>= traverse decodeHit)

readSqlite :: FilePath -> Text -> IO (Either DocumentationError (Maybe DocumentationRecord))
readSqlite target identifier = do
  pathExists <- doesFileExist (T.unpack identifier)
  absoluteIdentifier <- if pathExists then Just <$> makeAbsolute (T.unpack identifier) else pure Nothing
  let condition = case absoluteIdentifier of
        Just path -> "path = " <> sqlText (T.pack path)
        Nothing
          | "paclet:" `T.isPrefixOf` identifier ->
              "paclet = " <> sqlText identifier <> " COLLATE NOCASE"
          | otherwise ->
              "title = " <> sqlText identifier <> " COLLATE NOCASE OR paclet = " <> sqlText identifier <> " COLLATE NOCASE"
      statement =
        "SELECT id, title, paclet, kind, category, path, preview, text FROM documents WHERE "
          <> condition
          <> " LIMIT 1;"
  rows <- sqliteJsonQuery target statement
  pure $ do
    values <- rows
    case values of
      [] -> Right Nothing
      value : _ -> Just <$> decodeRecord value

sqliteJsonQuery :: FilePath -> Text -> IO (Either DocumentationError [JsonValue])
sqliteJsonQuery target statement = do
  sqlite <- findExecutable "sqlite3"
  case sqlite of
    Nothing -> pure (Left (DocumentationError "sqlite3 is required to query the documentation index"))
    Just executable -> do
      processResult <- try (readProcessWithExitCode executable ["-json", target, T.unpack statement] "")
      case (processResult :: Either IOException (ExitCode, String, String)) of
        Left exception -> pure (Left (DocumentationError (T.pack (show exception))))
        Right (ExitFailure code, _, errors) ->
          pure (Left (DocumentationError ("sqlite3 exited with code " <> T.pack (show code) <> ": " <> T.strip (T.pack errors))))
        Right (ExitSuccess, output, _)
          | null (dropWhile isSpace output) -> pure (Right [])
          | otherwise -> case parseJson (T.pack output) of
              Left jsonError -> pure (Left (DocumentationError (jsonErrorMessage jsonError)))
              Right (JsonArray rows) -> pure (Right rows)
              Right _ -> pure (Left (DocumentationError "sqlite3 did not return a JSON array"))

decodeRecord :: JsonValue -> Either DocumentationError DocumentationRecord
decodeRecord value = do
  fields <- objectFields value
  DocumentationRecord
    <$> optionalIntegerField "id" fields
    <*> textField "title" fields
    <*> textField "paclet" fields
    <*> textField "kind" fields
    <*> textField "category" fields
    <*> (T.unpack <$> textField "path" fields)
    <*> textField "preview" fields
    <*> textField "text" fields

optionalIntegerField :: Text -> Map.Map Text JsonValue -> Either DocumentationError (Maybe Integer)
optionalIntegerField key fields = case Map.lookup key fields of
  Nothing -> Right Nothing
  Just JsonNull -> Right Nothing
  Just (JsonNumber value) ->
    maybe (Left (DocumentationError ("invalid integer field: " <> key))) (Right . Just)
      (readMaybe (T.unpack value))
  Just _ -> Left (DocumentationError ("invalid integer field: " <> key))

decodeHit :: JsonValue -> Either DocumentationError DocumentationHit
decodeHit value = do
  fields <- objectFields value
  DocumentationHit
    <$> textField "title" fields
    <*> textField "paclet" fields
    <*> textField "kind" fields
    <*> textField "category" fields
    <*> (T.unpack <$> textField "path" fields)
    <*> textField "preview" fields
    <*> textField "snippet" fields
    <*> numberField "score" fields

objectFields :: JsonValue -> Either DocumentationError (Map.Map Text JsonValue)
objectFields (JsonObject fields) = Right fields
objectFields _ = Left (DocumentationError "documentation index row must be a JSON object")

textField :: Text -> Map.Map Text JsonValue -> Either DocumentationError Text
textField key fields = case Map.lookup key fields of
  Just (JsonString value) -> Right value
  _ -> Left (DocumentationError ("documentation index row is missing text field " <> key))

numberField :: Text -> Map.Map Text JsonValue -> Either DocumentationError Text
numberField key fields = case Map.lookup key fields of
  Just (JsonNumber value) -> Right value
  _ -> Left (DocumentationError ("documentation index row is missing numeric field " <> key))

documentationNotebookFiles :: [FilePath] -> IO [FilePath]
documentationNotebookFiles roots = concat <$> traverse walkRoot roots
 where
  walkRoot root = do
    exists <- doesDirectoryExist root
    if exists then walk root else pure []
  walk path = do
    names <- sort <$> listDirectory path
    nested <- forM names $ \name -> do
      let child = path </> name
      isDirectory <- doesDirectoryExist child
      if isDirectory
        then walk child
        else pure [child | map toLower (takeExtension child) == ".nb"]
    pure (concat nested)

findDocumentationPaths :: WolframInstallation -> Text -> Int -> IO [FilePath]
findDocumentationPaths installation identifier limit = case stemFromIdentifier identifier of
  Nothing -> pure []
  Just stem -> do
    paths <- documentationNotebookFiles (installationDocsRoots installation)
    pure
      ( take limit
          [ path
          | path <- paths
          , map toLower (takeBaseName path) == map toLower (T.unpack stem)
          ]
      )

stemFromIdentifier :: Text -> Maybe Text
stemFromIdentifier identifier =
  let candidate
        | "paclet:" `T.isPrefixOf` identifier = lastOrEmpty (T.splitOn "/" identifier)
        | otherwise = T.pack (takeBaseName (T.unpack identifier))
   in if T.null candidate || T.any (not . validStemCharacter) candidate
        then Nothing
        else Just candidate
 where
  validStemCharacter character = isAlphaNum character || character `elem` ("_.-" :: String)
  lastOrEmpty [] = ""
  lastOrEmpty values = last values

extractTitle :: Text -> FilePath -> Text
extractTitle source path = case T.breakOn "WindowTitle" source of
  (_, remainder) | T.null remainder -> T.pack (takeBaseName path)
  (_, remainder) ->
    let afterName = T.drop (T.length "WindowTitle") remainder
        afterArrow = T.stripStart (maybe afterName id (T.stripPrefix "->" (T.stripStart afterName)))
     in case T.uncons afterArrow of
          Just ('"', _) ->
            let end = skipWolframString afterArrow 0
             in parseWolframStringLiteral (T.take end afterArrow)
          _ ->
            let value = T.takeWhile validTitleCharacter afterArrow
             in if T.null value then T.pack (takeBaseName path) else value
 where
  validTitleCharacter character = isAlphaNum character || character `elem` ("`.$_-" :: String)

inferKindAndPaclet :: FilePath -> (Text, Text, Text)
inferKindAndPaclet path = case componentAfter "ReferencePages" components of
  Just category ->
    let pacletCategory = maybe ("ref/" <> T.toLower category) id (lookup category referenceCategories)
     in ("reference", category, "paclet:" <> pacletCategory <> "/" <> stem)
  Nothing -> case firstSection components of
    Just (section, category) -> (category, section, "paclet:" <> category <> "/" <> stem)
    Nothing -> ("document", "Other", "paclet:document/" <> stem)
 where
  components = map T.pack (splitDirectories path)
  stem = T.pack (takeBaseName path)
  componentAfter name values = do
    index <- findIndex (== name) values
    case drop (index + 1) values of
      value : _ -> Just value
      [] -> Nothing
  firstSection values = firstMatch sectionCategories
   where
    firstMatch [] = Nothing
    firstMatch ((section, category) : rest)
      | section `elem` values = Just (section, category)
      | otherwise = firstMatch rest

referenceCategories :: [(Text, Text)]
referenceCategories =
  [ ("Symbols", "ref")
  , ("Programs", "ref/program")
  , ("MenuItems", "ref/menuitem")
  , ("Characters", "ref/character")
  , ("Entities", "ref/entity")
  , ("Interpreters", "ref/interpreter")
  , ("FrontEndObjects", "ref/frontendobject")
  ]

sectionCategories :: [(Text, Text)]
sectionCategories =
  [ ("Guides", "guide")
  , ("Tutorials", "tutorial")
  , ("HowTos", "howto")
  , ("Workflows", "workflow")
  , ("WorkflowGuides", "workflowguide")
  , ("ExamplePages", "example")
  ]

extractStringLiterals :: Text -> [Text]
extractStringLiterals source = reverse (go 0 [])
 where
  sourceLength = T.length source
  go index output
    | index >= sourceLength = output
    | "(*" `T.isPrefixOf` T.drop index source = go (skipWolframComment source index) output
    | T.index source index == '"' =
        let nextIndex = skipWolframString source index
            literal = T.take (nextIndex - index) (T.drop index source)
         in go nextIndex (parseWolframStringLiteral literal : output)
    | otherwise = go (index + 1) output

filterUsefulStrings :: [Text] -> [Text]
filterUsefulStrings = filter useful . map T.strip
 where
  useful value =
    not (T.null value)
      && not (Set.member value noiseLiterals)
      && not (isUuid value)
      && not (isCompressedBlob value)

noiseLiterals :: Set.Set Text
noiseLiterals =
  Set.fromList
    [ "AnchorBar", "AnchorBarGrid", "Columns", "ExampleCount", "ExampleSection"
    , "LinkHand", "ObjectNameTranslation", "PacletNameCell", "PrimaryExamplesSection"
    , "Rows", "SeeAlsoRelated", "Spacer1"
    ]

isUuid :: Text -> Bool
isUuid value = case T.splitOn "-" value of
  [a, b, c, d, e] ->
    map T.length [a, b, c, d, e] == [8, 4, 4, 4, 12]
      && T.all isHexDigit (a <> b <> c <> d <> e)
  _ -> False

isCompressedBlob :: Text -> Bool
isCompressedBlob value =
  T.length value >= 200 && T.all (`elem` ("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=:._-" :: String)) value

collapseText :: Int -> Text -> Text
collapseText limit value
  | T.length collapsed <= limit = collapsed
  | limit <= 0 = ""
  | otherwise = T.stripEnd (T.take (limit - 1) collapsed) <> "…"
 where
  collapsed = T.unwords (T.words value)

buildMatchQuery :: Text -> Text
buildMatchQuery query = case wordsBy (not . matchCharacter) query of
  [] -> "\"" <> T.replace "\"" "\"\"" query <> "\""
  terms -> T.intercalate " AND " ["\"" <> term <> "\"*" | term <- terms]
 where
  matchCharacter character = isAlphaNum character || character `elem` ("_.:/-" :: String)
  wordsBy separator = filter (not . T.null) . T.split separator

sqlText :: Text -> Text
sqlText value = "CAST(X'" <> TE.decodeUtf8 (hexBytes (TE.encodeUtf8 value)) <> "' AS TEXT)"
 where
  hexBytes = BS.concatMap (TE.encodeUtf8 . byteHex)
  byteHex byte =
    let digits = showHex byte ""
     in T.pack (replicate (2 - length digits) '0' <> digits)

notFound :: Text -> Either DocumentationError value
notFound identifier =
  Left (DocumentationError ("No documentation page found for " <> identifier <> "."))
